klippy/webhooks.py

```diff
@@ -280,3 +280,136 @@ def get_status(self, eventtime=0.):
         return {
             "action_call_remote_method": self._action_call_remote_method
         }
+
+
+SUBSCRIPTION_REFRESH_TIME = .25
+
+class StatusHandler:
+    def __init__(self, webhooks):
+        self.printer = webhooks.printer
+        self.webhooks = webhooks
+        self.ready = False
+        self.reactor = self.printer.get_reactor()
+        self.available_objects = {}
+        self.subscriptions = {}
+        self.subscription_timer = self.reactor.register_timer(
+            self._batch_subscription_handler, self.reactor.NEVER)
+
+        # Register events
+        self.printer.register_event_handler(
+            "klippy:ready", self._handle_ready)
+        self.printer.register_event_handler(
+            "gcode:request_restart", self._handle_restart)
+
+        # Register webhooks
+        webhooks.register_endpoint(
+            "objects/list",
+            self._handle_object_request)
+        webhooks.register_endpoint(
+            "objects/status",
+            self._handle_status_request)
+        webhooks.register_endpoint(
+            "objects/subscription",
+            self._handle_subscription_request)
+
+    def _handle_ready(self):
+        eventtime = self.reactor.monotonic()
+        self.available_objects = {}
+        objs = self.printer.lookup_objects()
+        status_objs = {n: o for n, o in objs if hasattr(o, "get_status")}
+        for name, obj in status_objs.items():
+            attrs = obj.get_status(eventtime)
+            self.available_objects[name] = attrs.keys()
+        self.ready = True
+
+    def _handle_restart(self, eventtime):
+        self.ready = False
+        self.reactor.update_timer(self.subscription_timer, self.reactor.NEVER)
+
+    def _batch_subscription_handler(self, eventtime):
+        status = self._process_status_request(self.subscriptions)
+        self.webhooks.call_remote_method(
+            "process_status_update", status=status)
+        return eventtime + SUBSCRIPTION_REFRESH_TIME
+
+    def _process_status_request(self, requested_objects):
+        result = {}
+        if self.ready:
+            for name, req_items in requested_objects.items():
+                obj = self.printer.lookup_object(name, None)
+                if obj is not None and name in self.available_objects:
+                    status = obj.get_status(self.reactor.monotonic())
```

Comment:
FYI, the call to `self.reactor.monotonic()` should be done outside the loop - for two reasons: the monotonic() call is a bit heavy (it requires a system call), and if two different `get_status()` calls are made with different times, there is the possibility that a status query spanning multiple objects may create an overall status that isn't internally in sync.

Reply:
Good call.  I'll fix this up.